### PR TITLE
distsql: avoid planning on a draining node.

### LINKED ISSUE #100578: sql: distsql planning should avoid sending flows to SQL servers in the process of starting up, and in the process of shutting down
New description: we would like a new field in the sql instances table that indicates node readiness.

This would not be set to "ready" before server startup has completed; and it would also be set to a different value during graceful shutdown.

Distsql planning should avoid nodes based on that field.

----

Original title: **jobs,upgrades: deadlock during migrations when initializing multiple servers for 1 tenant**

**Describe the problem**

While investigating #100436 I discovered the following deadlock:

1. create a new tenant record; no server running yet. At this point the tenant keyspace is empty *and it has not run its migrations yet*
2. start **two or more servers for that tenant simultaneously**  (this is important)
3. observe: often, the tenant servers deadlock.

The underlying problem is the following:

- when the tenant servers **begin** their startup, they immediately populate a SQL instance row in the liveness table. **However at that time they are not yet ready to process SQL queries because their migrations have not run yet**
- then, one of the two will start running the upgrade migrations.
- one of these migrations will need a distsql distributed query. That, in turn, will fail because the other server(s) are not yet ready to accept distsql execution:

```
I230404 10:23:32.796367 15064 upgrade/upgrademanager/manager.go:235 ⋮ [T2,n2] 534  the last permanent upgrade (v1000022.2-94) does not appear to have completed; attempting to run all upgrades
I230404 10:23:32.796659 15064 upgrade/upgrademanager/manager.go:278 ⋮ [T2,n2] 535  running permanent upgrade for version 0.0-2
I230404 10:23:32.824530 15776 1@circuitbreaker/circuitbreaker.go:322 ⋮ [T2,n2] 545  circuitbreaker: ‹rpc 127.0.0.1:64403 [n1]› tripped: unable to look up descriptor for n1: non existent SQL instance
I230404 10:23:32.824557 15776 1@circuitbreaker/circuitbreaker.go:447 ⋮ [T2,n2] 546  circuitbreaker: ‹rpc 127.0.0.1:64403 [n1]› event: ‹BreakerTripped›
W230404 10:23:32.824573 15776 sql/colflow/colrpc/outbox.go:189 ⋮ [T2,n2,intExec=‹select-job›,f‹0f758d2f›,distsql.stmt=‹WITH latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE (info_key = '_') AND (job_id = $1) ORDER BY written DESC LIMIT _), latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WH
ERE (info_key = '_') AND (job_id = $1) ORDER BY written DESC LIMIT _) SELECT status, payload.value AS payload, progress.value AS progress, claim_session_id, COALESCE(last_run, created), COALESCE(num_runs, _) FROM system.jobs AS j INNER JOIN latestpayload AS payload ON j.id =›,distsql.gateway=‹2›,distsql.appname=‹$ internal-select-job›,distsq
l.txn=‹b8a5af83-f1f1-4364-9074-5e4bc41d66f5›,streamID=‹9›] 547  Outbox Dial connection error, distributed query will fail: unable to look up descriptor for n1: non existent SQL instance
```

**Expected behavior**

DistSQL should recognize when other SQL instances are not yet ready to process queries and not attempt to use them.

Alternatively we could disable DistSQL entirely during upgrade migrations; but that seems undesirable because certain migrations rewrite tables and that would benefit from the extra concurrency.

Jira issue: CRDB-26509

Epic CRDB-39091

### PR DESCRIPTION
**Pull request by shubhamdhama**
Commit-wise description

---
**Commit 1: [`sql: add is_draining to system.sql_instances table.`](https://github.com/cockroachdb/cockroach/pull/128283/commits/4b0de41bc39)**

This is a preliminary commit toward improving the DistSQL to avoid planning
on an instance that is draining. The idea here is to set this column to
true as soon as the draining request is received by the server and use that
through instance reader in the planner at gateway.

Adding this column requires writing a migration for the upgrade. And since
this new column is nullable we don't need to backfill and we are good with
just swapping the descriptors.

Informs: #100578
Release note: None

Development notes:

To update `pkg/sql/catalog/bootstrap/testdata/testdata` I had to manually
update the hash values for `system`/`tenant` in this file and pass the
`--rewrite` option to do the rest:
```
dev test //pkg/sql/catalog/bootstrap:bootstrap_test \
        --filter=TestInitialValuesToString/testdata --rewrite
```

To update tests in
`pkg/sql/catalog/systemschema_test/testdata/bootstrap_{system,tenant}`
```
dev test \
        //pkg/sql/catalog/systemschema_test:systemschema_test_test --rewrite
```

---
**Commit 2: [`distsql: avoid planning on a draining node.`](https://github.com/cockroachdb/cockroach/pull/128283/commits/6337d932854)**

This change would prevent distsql planner to include a draining node in a
plan by leveraging the system.sql_instances table's `is_draining` field.

Somewhat related work to avoid planning on an unavailable node: #124986

Fixes: #100578
Epic: CRDB-31575
Release note: None

### CODE REVIEW

**Comment by shubhamdhama** [comment]
*[`pkg/sql/distsql_physical_planner_test.go` line 832 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O3TrRoG30ni_-oy3qhA:-O3TrRoG30ni_-oy3qhB:b11sc2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc10bf170e45c012dbb78c228457d8f3c721b12b/pkg/sql/distsql_physical_planner_test.go#L832)):*
> ```Go
> 
> 			return checkQueryPlan(t, r, expectedPlans)
> 		}, 4*time.Second)
> ```

The test for external-process mode is failing because the plan still includes the drained 5th node. This happens because setting `is_draining` to true isn't being reflected on the gateway node. However, this works correctly on a roachprod cluster. Despite extensive investigation, I couldn't figure out why this is happening, so I would really appreciate any help.



_Code quote:_
```Go
		testutils.SucceedsWithin(t, func() error {
			// TODO: REMOVE
			res := r.QueryStr(t, "SELECT now(), id, encode(session_id, 'hex'), is_draining FROM system.sql_instances WHERE addr IS NOT NULL")
			t.Logf("res : %v", res)

			return checkQueryPlan(t, r, expectedPlans)
		}, 4*time.Second)
```

File: pkg/sql/distsql_physical_planner.go [hunk 1]
```diff
@@ -1593,6 +1593,24 @@ func (dsp *DistSQLPlanner) checkInstanceHealth(
...
+		return nil, err
+	}
+
+	availableInstances := make([]sqlinstance.InstanceInfo, 0, len(instances))
+	for _, n := range instances {
+		if n.IsDraining {
+			continue
+		}
+		availableInstances = append(availableInstances, n)
+	}
```
Comment (shubhamdhama) [inline pkg/sql/distsql_physical_planner.go:1610]: Change to in-place filtering

**Review by shubhamdhama — COMMENTED** [review]
This is also ready for review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-:-O5Ho_WmEMFupaygNEfB:bvgi8hu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @jeffswenson, @stevendanna, and @yuzefovich)

___
*[`pkg/sql/distsql_physical_planner_test.go` line 832 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O3TrRoG30ni_-oy3qhA:-O5HoIYI2LCPU_FwcQxS:brsqksm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc10bf170e45c012dbb78c228457d8f3c721b12b/pkg/sql/distsql_physical_planner_test.go#L832)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

The test for external-process mode is failing because the plan still includes the drained 5th node. This happens because setting `is_draining` to true isn't being reflected on the gateway node. However, this works correctly on a roachprod cluster. Despite extensive investigation, I couldn't figure out why this is happening, so I would really appreciate any help.
</blockquote></details>

Oh btw, this was fixed. Earlier, I was draining system tenant instance which I now changed with the correct sql instance.

**Review by jeffswenson — APPROVED** [review]
LGTM

I think you should also get a review from someone on sql foundations or sql queries. The functionality straddles the two teams. Historically sql foundations owned the sql_instances infrastructure and sql queries still owns distsql planning which is impacted by this change.

File: pkg/sql/sqlinstance/instancestorage/test_helpers.go [hunk 2]
```diff
@@ -153,3 +160,10 @@ func (s *Storage) GetAllInstancesDataForTest(
 	}
 	return makeInstanceInfos(rows), nil
 }
+
+// SortInstaces sorts instances by their id.
+func SortInstances(instances []sqlinstance.InstanceInfo) {
+	sort.SliceStable(instances, func(idx1, idx2 int) bool {
```
[the remark this answers is not in the document]
Comment (shubhamdhama) [reply pkg/sql/sqlinstance/instancestorage/test_helpers.go]: Right, I blindly moved this from other location to here. Done!

File: pkg/sql/distsql_physical_planner.go [hunk 3]
```diff
@@ -1593,6 +1593,24 @@ func (dsp *DistSQLPlanner) checkInstanceHealth(
 	return status
 }
 
+func (dsp *DistSQLPlanner) GetAvailableInstances(
+	ctx context.Context,
+) ([]sqlinstance.InstanceInfo, error) {
+	instances, err := dsp.sqlAddressResolver.GetAllInstances(ctx)
```
[the remark this answers is not in the document]
Comment (shubhamdhama) [reply pkg/sql/distsql_physical_planner.go]: Make sense, it does reduce some more line of code. Done!

File: pkg/server/drain.go [hunk 4]
```diff
@@ -332,6 +332,17 @@ func (s *drainServer) runDrain(
 func (s *drainServer) drainInner(
 	ctx context.Context, reporter func(int, redact.SafeString), verbose bool,
 ) (err error) {
+	// TODO: REMOVE
```
[the remark this answers is not in the document]
Comment (shubhamdhama) [reply pkg/server/drain.go]: Yes, thanks!

File: pkg/sql/distsql_physical_planner.go (hunk 1 above)
Comment (shubhamdhama) [reply pkg/sql/distsql_physical_planner.go]: Removed.

**Review by DrewKimball — APPROVED** [review]
:lgtm: I just have a few questions.

> Reviewed 6 of 22 files at r1, 6 of 18 files at r3, 13 of 13 files at r5, 4 of 4 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-:-O5VC02tCr_v48WeCoz7:beasevm)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @jeffswenson, @shubhamdhama, and @stevendanna)

___
*[`pkg/sql/distsql_physical_planner.go` line 1725 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5VBi9r6qobBXhf8U2h:-O5VBi9r6qobBXhf8U2i:b-rgxxjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f8894ce28bb4be2b137750d3587322e76d33f43/pkg/sql/distsql_physical_planner.go#L1725)):*
> ```Go
> 	}
> 
> 	// GetAvailableInstances returns mostly healthy nodes, except those that have
> ```

[nit] Why the comment change?
___
*[`pkg/sql/sqlinstance/instancestorage/instancereader.go` line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5SIfJP5IrGP4-j1Bxt:-O5SIfJP5IrGP4-j1Bxu:b-aofos1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbd9169dbf1fa83b6fac4ae2289c30cb13c0fe45/pkg/sql/sqlinstance/instancestorage/instancereader.go#L138)):*
> ```Go
> 		Locality:        row.locality,
> 		BinaryVersion:   row.binaryVersion,
> 		Region:          row.region,
> ```

Is it problematic that we weren't filling in `Region` before?
___
*[`pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go` line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5VAZRoEb50Kgc0rNe3:-O5VAZRoEb50Kgc0rNe4:b-pvrhup) ([raw file](https://github.com/cockroachdb/cockroach/blob/007f193c937ef1471ae819368b1f92b37e670f1d/pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go#L39)):*
> ```Go
> 		mutableDesc.TableDescriptor = *expectedDesc
> 		mutableDesc.Version = version
> 		return txn.Descriptors().WriteDesc(ctx, false, mutableDesc, txn.KV())
> ```

Will `WriteDesc` increment the descriptor version?

**Review by shubhamdhama — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128283)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @dt, @jeffswenson, and @stevendanna)

___
*[`pkg/sql/distsql_physical_planner.go` line 1725 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5VBi9r6qobBXhf8U2h:-O5lH80T3MjVB0JYuVpJ:b-aojsqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f8894ce28bb4be2b137750d3587322e76d33f43/pkg/sql/distsql_physical_planner.go#L1725)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Why the comment change?
</blockquote></details>

Done. (Nice catch, in my previous revision I removed this function).
___
*[`pkg/sql/sqlinstance/instancestorage/instancereader.go` line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5SIfJP5IrGP4-j1Bxt:-O5lHFysDpTabUF-qWtm:b2g43ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbd9169dbf1fa83b6fac4ae2289c30cb13c0fe45/pkg/sql/sqlinstance/instancestorage/instancereader.go#L138)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is it problematic that we weren't filling in `Region` before?
</blockquote></details>

It wasn't problematic earlier, but I figured it's most likely a miss.
___
*[`pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go` line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5VAZRoEb50Kgc0rNe3:-O5lIE930DMHMEtha-vK:bsl84b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/007f193c937ef1471ae819368b1f92b37e670f1d/pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go#L39)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Will `WriteDesc` increment the descriptor version?
</blockquote></details>

Yes,

https://github.com/cockroachdb/cockroach/blob/8551145a0c99c4c95a28ec470e699d0c20ca97ab/pkg/sql/catalog/descs/collection.go#L559-L566

https://github.com/cockroachdb/cockroach/blob/8551145a0c99c4c95a28ec470e699d0c20ca97ab/pkg/sql/catalog/descs/collection.go#L276-L289

**Comment by rimadeodhar** [comment]
@fqazi: Please could you take a look from the SQL Foundations side? Thanks.

**Review by fqazi — APPROVED** [review]
:lgtm: 

One minor nit

> Reviewed 6 of 22 files at r1, 4 of 18 files at r3, 3 of 13 files at r5, 2 of 4 files at r6, 10 of 10 files at r8, 4 of 4 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-:-O5srnfY0sAnOy__KhDk:b-r5uyez)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @dt, @jeffswenson, @shubhamdhama, and @stevendanna)

___
*[`pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5spid8239STErCobPq:-O5spid9EpDaAvz0Y6HE:b-85od4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b0de41bc395338d47de76cd836f90df06d2e0e2/pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go#L37)):*
> ```Go
> 		}
> 		version := mutableDesc.Version
> 		mutableDesc.TableDescriptor = *expectedDesc
> ```

Use protoutil.Clone here

**Comment by shubhamdhama** [comment]
*[`pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/128283#-O5spid8239STErCobPq:-O5svK0ZBAjdbmnhBLyM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b0de41bc395338d47de76cd836f90df06d2e0e2/pkg/upgrade/upgrades/v24_3_sql_instances_add_draining.go#L37)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Use protoutil.Clone here
</blockquote></details>

Done.

File: pkg/sql/sqlinstance/instancestorage/test_helpers.go [hunk 5]
```diff
@@ -153,3 +160,10 @@ func (s *Storage) GetAllInstancesDataForTest(
 	}
 	return makeInstanceInfos(rows), nil
 }
+
+// SortInstaces sorts instances by their id.
```
Comment (stevendanna) [inline pkg/sql/sqlinstance/instancestorage/test_helpers.go:164]: nit: SortInstances

**Comment by shubhamdhama** [comment]
Thanks everyone for the review.

bors r=jeffswenson,DrewKimball,fqazi

File: pkg/upgrade/upgrades/v24_3_sql_instances_add_draining_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		const firstId = base.SQLInstanceID(1000)
+		createInstance(firstId, true, true)
+		const secondId = base.SQLInstanceID(1001)
+		createInstance(secondId, true, false)
+		const thirdId = base.SQLInstanceID(1002)
+		createInstance(thirdId, false, false)
+
+		// Here we show old readers (SQL, instancereader) doesn't break if
+		// they find new schema value before cluster upgrade. This gives extra
+		// confidence on top of version guard we already have.
```
Comment (stevendanna) [inline pkg/upgrade/upgrades/v24_3_sql_instances_add_draining_test.go:178]: nit: doesn't break -> don't break

One note is that here by "old readers" we mean readers on the old _cluster version_. But it doesn't test readers on the old binary version.

File: pkg/sql/distsql_physical_planner_test.go [hunk 7]
```diff
@@ -664,88 +675,166 @@ func TestDistSQLUnavailableHosts(t *testing.T) {
...
-		numNodes,
-		base.TestClusterArgs{
-			ReplicationMode: base.ReplicationManual,
-			ServerArgs:      base.TestServerArgs{Knobs: base.TestingKnobs{DistSQL: &execinfra.TestingKnobs{DrainFast: true}}, UseDatabase: "test"},
-		},
-	)
-	ctx := context.Background()
-	defer tc.Stopper().Stop(ctx)
+	t.Run("draining-nodes-in-single-tenant-mode", func(t *testing.T) {
+		skip.UnderDuress(t, "takes 20s")
```
Comment (stevendanna) [inline pkg/sql/distsql_physical_planner_test.go:699]: I think you can move this skip to the overall test rather that putting it in each subtest.

**Review by stevendanna — APPROVED** [review]
Thanks for working on this. Overall this looks reasonable to me.

**Comment by shubhamdhama** [comment]
TFTR.

bors r=jeffswenson,DrewKimball,fqazi,stevendanna

**Pull request merged**